File: src/lua/test_runner.lua
--[[
test_runner.lua — Minimal Lua test framework for Zay plugins.

Provides describe/it/assert API similar to Busted or LuaUnit.

The module is pre-set as a global (`test_runner`) by the Zig runner
(test_runner.zig), NOT loaded via `require("test_runner")`. Test files may
call `test.run()` explicitly (readable when run standalone), but the Zig
runner also auto-runs it after loading the file, so an explicit call is
optional and never double-executes (run() is idempotent).

Usage:
  local test = test_runner
  test.describe("my plugin", function()
    test.it("does something", function()
      test.assert.equal(4, 2 + 2)
    end)
  end)
  test.run()
]]

local test_runner = {}

-- Test state
local suites = {}
local current_suite = nil
local failures = 0
local successes = 0

-- ANSI color codes (when available)
local color = {
  green = "\27[32m",
  red = "\27[31m",
  yellow = "\27[33m",
  cyan = "\27[36m",
  reset = "\27[0m",
}

--- Define a test suite.
-- @param name string: suite description
-- @param fn function: suite body containing it() calls
function test_runner.describe(name, fn)
  table.insert(suites, { name = name, tests = {} })
  current_suite = suites[#suites]
  local ok, err = pcall(fn)
  if not ok then
    io.stderr:write(string.format("SUITE ERROR: %s\n", err))
  end
  current_suite = nil
end

--- Define a single test case.
-- @param name string: test description
-- @param fn function: test body; raises an error on failure
function test_runner.it(name, fn)
  if not current_suite then
    error("it() must be called inside describe()")
  end
  table.insert(current_suite.tests, { name = name, fn = fn })
end

--- Assertion functions.
test_runner.assert = {}

--- Assert that a value is truthy.
function test_runner.assert.is_true(value, message)
  if not value then
    error(message or "expected truthy value, got " .. tostring(value))
  end
end

--- Assert that a value is falsy.
function test_runner.assert.is_false(value, message)
  if value then
    error(message or "expected falsy value, got " .. tostring(value))
  end
end

--- Assert that two values are equal (using ==).
function test_runner.assert.equal(expected, actual, message)
  if expected ~= actual then
    error(message or string.format("expected %s, got %s", tostring(expected), tostring(actual)))
  end
end

--- Assert that two values are NOT equal.
function test_runner.assert.not_equal(a, b, message)
  if a == b then
    error(message or string.format("expected %s ~= %s", tostring(a), tostring(b)))
  end
end

--- Assert that a string matches a pattern.
function test_runner.assert.matches(pattern, str, message)
  if not string.find(str, pattern) then
    error(message or string.format("expected '%s' to match '%s'", str, pattern))
  end
end

--- Assert that a function raises an error.
function test_runner.assert.error(fn, expected_message, message)
  local ok, err = pcall(fn)
  if ok then
    error(message or "expected error, but none was raised")
  end
  if expected_message and not string.find(err, expected_message) then
    error(message or string.format("expected error containing '%s', got '%s'", expected_message, err))
  end
end

--- Assert that a table has a given key.
function test_runner.assert.has_key(key, tbl, message)
  if tbl[key] == nil then
    error(message or string.format("expected table to have key '%s'", key))
  end
end

--- Assert that a string contains a substring.
function test_runner.assert.contains(sub, str, message)
  if not string.find(str, sub, 1, true) then
    error(message or string.format("expected '%s' to contain '%s'", str, sub))
  end
end

--- Run all registered test suites and print results.
-- Idempotent: the first call executes the suites, clears them, and caches the
-- verdict; a second call (e.g. the Zig runner auto-running after a file that
-- already called run()) returns the cached verdict without re-executing. This
-- lets the Zig runner always call run() after loading a file without
-- double-executing files that call it themselves.
-- @return boolean: true if all tests passed. False when there are zero tests
--   (a test file with no `it` blocks is a failure — it proves nothing).
function test_runner.run()
  if test_runner._has_run then
    return test_runner._last_result
  end
  test_runner._has_run = true

  local suites_snapshot = suites
  suites = {}
  current_suite = nil

  local total = 0
  local passed = 0
  local failed = 0

  io.write(color.cyan .. "Running Lua plugin tests...\n" .. color.reset)

  for _, suite in ipairs(suites_snapshot) do
    io.write(string.format("  %s:\n", suite.name))
    for _, test in ipairs(suite.tests) do
      total = total + 1
      local ok, err = pcall(test.fn)
      if ok then
        passed = passed + 1
        io.write(string.format("    %s✓%s %s\n", color.green, color.reset, test.name))
      else
        failed = failed + 1
        io.write(string.format("    %s✗%s %s\n", color.red, color.reset, test.name))
        io.write(string.format("      %s%s%s\n", color.yellow, err, color.reset))
      end
    end
  end

  io.write(string.format("\n%d tests: %d passed, %d failed\n", total, passed, failed))

  local result
  if total == 0 then
    io.write(color.red .. "NO TESTS FOUND — a test file with no `it` blocks is a failure\n" .. color.reset)
    result = false
  elseif failed > 0 then
    io.write(color.red .. "SOME TESTS FAILED\n" .. color.reset)
    result = false
  else
    io.write(color.green .. "ALL TESTS PASSED\n" .. color.reset)
    result = true
  end

  test_runner._last_result = result
  return result
end

return test_runner

File: src/lua/test_runner.zig
//! Lua test runner for Zay plugins.
//!
//! Loads the test_runner.lua library, then loads and runs Lua test files.
//! Each test file is executed in a sandboxed Lua state with the test_runner
//! module pre-set as the global `test_runner` (not via `require`). After the
//! file chunk runs, `test_runner.run()` is auto-invoked and its boolean
//! verdict is the pass/fail gate — a file with zero tests fails.

const std = @import("std");
const log = std.log.scoped(.lua);
const c = @import("c");
const State = @import("../lua/state.zig").State;
const sandbox = @import("../lua/sandbox.zig");

/// Run Lua test files. Each file is loaded into a fresh sandboxed state
/// with the test_runner module pre-loaded. Returns true if all tests pass.
pub fn runTestFiles(gpa: std.mem.Allocator, io: std.Io, file_paths: []const []const u8) !bool {
    var all_passed = true;
    for (file_paths) |path| {
        const passed = try runSingleTestFile(gpa, io, path);
        if (!passed) all_passed = false;
    }
    return all_passed;
}

/// Run a single Lua test file.
fn runSingleTestFile(gpa: std.mem.Allocator, io: std.Io, path: []const u8) !bool {
    // Read the test file
    const content = try readFile(gpa, io, path);
    defer gpa.free(content);
    return runTestSource(gpa, io, content);
}

/// Run a Lua test file's source in a fresh sandboxed state and report whether
/// every assertion passed. The test_runner module is pre-loaded as a global;
/// the file chunk is executed, then `test_runner.run()` is auto-invoked and its
/// boolean result is the pass/fail verdict. A file passes iff the chunk runs
/// cleanly AND run() returns true (which is false for zero tests).
fn runTestSource(gpa: std.mem.Allocator, io: std.Io, content: []const u8) !bool {
    var L = try sandbox.createSandboxedStateWithIo(.{ .full_access = true }, io);
    defer {
        sandbox.freeHookData(L.handle);
        L.deinit();
    }

    // Pre-load the test_runner module
    const runner_src = @embedFile("test_runner.lua");
    const null_term_runner = try std.fmt.allocPrintSentinel(gpa, "{s}", .{runner_src}, 0);
    defer gpa.free(null_term_runner);

    // Load and execute the test runner library
    if (!L.doString(null_term_runner)) {
        const err = L.getErrorMessage();
        log.warn("lua.test_runner.load_failed err={s}", .{err orelse "unknown"});
        return false;
    }

    // The test_runner module is now in the global table as the return value.
    // Store it as a global so test files can reference it.
    _ = c.lua_setglobal(L.handle, "test_runner");

    // Load and execute the test file
    const null_term_test = try std.fmt.allocPrintSentinel(gpa, "{s}", .{content}, 0);
    defer gpa.free(null_term_test);

    // Load the test file as a function
    L.loadString(null_term_test) catch {
        const err = L.getErrorMessage();
        log.warn("lua.test_file.load_failed err={s}", .{err orelse "unknown"});
        L.pop(1);
        return false;
    };

    // Call it — the test file registers suites (it may also call test.run()
    // itself; run() is idempotent so a later auto-run is a no-op).
    const rc = L.pcall(0, 0);
    if (rc != c.LUA_OK) {
        const err = L.getErrorMessage();
        log.warn("lua.test_file.run_failed err={s}", .{err orelse "unknown"});
        L.pop(1);
        return false;
    }

    // Auto-run: load `return test_runner.run()` and read the boolean verdict.
    // This is load-bearing — a file that never calls test.run() still gets its
    // suites executed and its pass/fail propagated, and an empty file (zero
    // `it` blocks) fails.
    const auto_run = "return test_runner.run()";
    L.loadString(auto_run) catch {
        const err = L.getErrorMessage();
        log.warn("lua.test_runner.autorun_load_failed err={s}", .{err orelse "unknown"});
        L.pop(1);
        return false;
    };
    const run_rc = L.pcall(0, 1);
    if (run_rc != c.LUA_OK) {
        const err = L.getErrorMessage();
        log.warn("lua.test_runner.autorun_failed err={s}", .{err orelse "unknown"});
        L.pop(1);
        return false;
    }
    const passed = L.toBoolean(-1);
    L.pop(1);
    return passed;
}

/// Read a file's contents into an owned slice.
fn readFile(gpa: std.mem.Allocator, io: std.Io, path: []const u8) ![]u8 {
    var file = try std.Io.Dir.openFile(.cwd(), io, path, .{});
    defer file.close(io);
    const stat = try file.stat(io);
    const bytes = try gpa.alloc(u8, @intCast(stat.size));
    errdefer gpa.free(bytes);
    var reader = file.reader(io, &.{});
    try reader.interface.readSliceAll(bytes);
    return bytes;
}

// ── Tests ────────────────────────────────────────────────────────────
//
// runTestSource is exercised directly with source slices so the tests don't
// touch the filesystem. The state is created without Io (full-access sandbox),
// so it can run inline.

test "test runner: passing assertion reports true" {
    const src =
        \\local test = test_runner
        \\test.describe("s", function()
        \\  test.it("passes", function() test.assert.equal(4, 2 + 2) end)
        \\end)
    ;
    try std.testing.expect(try runTestSource(std.testing.allocator, std.testing.io, src));
}

test "test runner: failing assertion reports false (the regression)" {
    const src =
        \\local test = test_runner
        \\test.describe("s", function()
        \\  test.it("fails", function() test.assert.equal(4, 5) end)
        \\end)
    ;
    try std.testing.expect(!(try runTestSource(std.testing.allocator, std.testing.io, src)));
}

test "test runner: file that never calls test.run() still reports correctly" {
    const src =
        \\local test = test_runner
        \\test.describe("s", function()
        \\  test.it("passes", function() test.assert.equal(1, 1) end)
        \\end)
        \\-- no explicit test.run() — the Zig runner auto-runs it
    ;
    try std.testing.expect(try runTestSource(std.testing.allocator, std.testing.io, src));
}

test "test runner: empty file (zero it blocks) reports false" {
    const src =
        \\local test = test_runner
        \\test.describe("s", function()
        \\  -- no tests
        \\end)
    ;
    try std.testing.expect(!(try runTestSource(std.testing.allocator, std.testing.io, src)));
}

test "test runner: explicit test.run() then auto-run does not double-execute" {
    const src =
        \\local test = test_runner
        \\test.describe("s", function()
        \\  test.it("passes", function() test.assert.equal(2, 2) end)
        \\end)
        \\test.run()
    ;
    try std.testing.expect(try runTestSource(std.testing.allocator, std.testing.io, src));
}

test "test runner: syntax error in file reports false" {
    const src = "this is not lua @@@";
    try std.testing.expect(!(try runTestSource(std.testing.allocator, std.testing.io, src)));
}
